### PR TITLE
Implement PolarizationAngleCorrectionMD (master)

### LINKED ISSUE #31679: Implement algorithm PolarizationAngleCorrectionMD
### Expected behavior

This is ORNL [SPEC #96](https://code.ornl.gov/sns-hfir-scse/spectroscopy/spectroscopy/-/issues/96)

This is equivalent to HyspecScharpfCorrection

workspace must be in Q_lab or Q_sample. |Q| is not good

user input is an in-plane polarization angle PAP_APA​, between -180 and 180 degrees, and a precision (between 0 and 1)
for every event one needs to calculate the horizontal plane angle γ\gammaγ between momentum transfer and direct beam. Here are equivalent options:

if the dimensions are in Q_lab
γ=arctan⁡2(Qx,Qz)\gamma = \arctan2(Q_x,Q_z)γ=arctan2(Qx​,Qz​)

if the dimensions are in Q_sample, convert first to Q_lab. Get the goniometer rotation matrix RRR from the corresponding experiment info and use Qlab=RQsampleQ_{lab}=RQ_{sample}Qlab​=RQsample​

in both cases one can instead calculate kf/kik_f/k_ikf​/ki​ from incident energy, energy transfer, and use detector position (θ\thetaθ) to calculate the angle:
kfki=Ei−ΔEEiγ=arctan⁡2(−kfkisin⁡θ,1−kfkicos⁡θ)\frac{k_f}{k_i}=\sqrt{\frac{E_i-\Delta E}{E_i}}\\
\gamma=\arctan2(-\frac{k_f}{k_i}\sin\theta,1-\frac{k_f}{k_i}\cos\theta)ki​kf​​=Ei​Ei​−ΔE​​γ=arctan2(−ki​kf​​sinθ,1−ki​kf​​cosθ)

The Scharpf angle α=γ−PA\alpha=\gamma-P_Aα=γ−PA​

Calculate cos⁡(2α)\cos(2\alpha)cos(2α) for every event. If greater in absolute value than the precision, then
F=1cos⁡(2α)F=\frac 1{\cos(2\alpha)}F=cos(2α)1​
else F=0F=0F=0

For every event
I∗=FErr2∗=F2I *= F\\
Err^2 *= F^2I∗=FErr2∗=F2

To test
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

we1 = CreateSampleWorkspace(WorkspaceType='Event',
                           Function='Flat background',
                           BankPixelWidth=1,
                           XUnit='DeltaE',
                           XMin=-10, 
                           XMax=19, 
                           BinWidth=0.5)
AddSampleLog(Workspace=we1,LogName='Ei', LogText='20.', LogType='Number')
MoveInstrumentComponent(Workspace=we1, ComponentName='bank1', X=3, Z=3, RelativePosition=False)
MoveInstrumentComponent(Workspace=we1, ComponentName='bank2', X=-3, Z=-3, RelativePosition=False)
we2=CloneWorkspace(we1)
SetGoniometer(Workspace=we1, Axis0='0,0,1,0,1')
SetGoniometer(Workspace=we2, Axis0='30,0,1,0,1')

# old way
we1c = HyspecScharpfCorrection(InputWorkspace=we1,
                               PolarizationAngle=-10,
                               Precision=0.2)
md1c = ConvertToMD(InputWorkspace=we1c, QDimensions='Q3D')
we2c = HyspecScharpfCorrection(InputWorkspace=we2,
                               PolarizationAngle=-10,
                               Precision=0.2)
md2c = ConvertToMD(InputWorkspace=we2c, QDimensions='Q3D')
mdpac = MergeMD(InputWorkspaces='md1c, md2c')

# new way
md1 = ConvertToMD(InputWorkspace=we1, QDimensions='Q3D')
md2 = ConvertToMD(InputWorkspace=we2, QDimensions='Q3D')
md = MergeMD(InputWorkspaces='md1,md2')
# mdpacMD = PolarizationAngleCorrectionMD(InputWorkspace=md, PolarizationAngle=-10, Precision=0.2)
```

### PR DESCRIPTION
**Pull request by wdzhou**
**Description of work.**

PolarizationAngleCorrectionMD

**To test:**

```
       # import mantid algorithms, numpy and matplotlib
       from mantid.simpleapi import *
       import matplotlib.pyplot as plt
       import numpy as np

       we1 = CreateSampleWorkspace(WorkspaceType='Event',
                                  Function='Flat background',
                                  BankPixelWidth=1,
                                  XUnit='DeltaE',
                                  XMin=-10,
                                  XMax=19,
                                  BinWidth=0.5)
       AddSampleLog(Workspace=we1,LogName='Ei', LogText='20.', LogType='Number')
       MoveInstrumentComponent(Workspace=we1, ComponentName='bank1', X=3, Z=3, RelativePosition=False)
       MoveInstrumentComponent(Workspace=we1, ComponentName='bank2', X=-3, Z=-3, RelativePosition=False)
       we2=CloneWorkspace(we1)
       SetGoniometer(Workspace=we1, Axis0='0,0,1,0,1')
       SetGoniometer(Workspace=we2, Axis0='30,0,1,0,1')

       # old way
       we1c = HyspecScharpfCorrection(InputWorkspace=we1,
                                      PolarizationAngle=-10,
                                      Precision=0.2)
       md1c = ConvertToMD(InputWorkspace=we1c, QDimensions='Q3D')
       we2c = HyspecScharpfCorrection(InputWorkspace=we2,
                                      PolarizationAngle=-10,
                                      Precision=0.2)
       md2c = ConvertToMD(InputWorkspace=we2c, QDimensions='Q3D')
       mdpac = MergeMD(InputWorkspaces='md1c, md2c')

       # new way
       md1 = ConvertToMD(InputWorkspace=we1, QDimensions='Q3D')
       md2 = ConvertToMD(InputWorkspace=we2, QDimensions='Q3D')
       md = MergeMD(InputWorkspaces='md1,md2')
       mdpacMD = PolarizationAngleCorrectionMD(InputWorkspace=md, PolarizationAngle=-10, Precision=0.2)
```

Fixes #31679. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/PolarizationAngleCorrectionMD-v1.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+Inputs
+======
+
+Input MDEventWorkspace must be in Q-lab or Q-sample. :math:`|Q|` is not allowed.
+
+Usage
+-----
+
+**Example - Run Applied Detailed Balance**
```
Comment (AndreiSavici) [inline docs/source/algorithms/PolarizationAngleCorrectionMD-v1.rst:40]: Wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+    std::string workspaces(corrected_qsample_name + ", " + temp_md2);
+    MergeMD merge_alg;
+    merge_alg.initialize();
+    merge_alg.setProperty("InputWorkspaces", workspaces);
+    merge_alg.setProperty("OutputWorkspace", corrected_qsample_merged_name);
+    merge_alg.execute();
+  }
+
+  /**
+   * @brief Apply detailed balance to event workspace
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:380]: Wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    Mantid::DataHandling::SaveNexusProcessed save;
+    save.initialize();
+    save.setPropertyValue("InputWorkspace", temp_event_ws0);
+    save.setPropertyValue("Filename", "/tmp/gold0.nxs");
+    save.execute();
+
+    convertToMD(temp_event_ws0, corrected_qsample_name, "Q3D", "Q_sample");
+    convertToMD(temp_event_ws0, corrected_qlab_name, "Q3D", "Q_lab");
+
+    // apply detailed balance and convert to MD for event workspace 2
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:364]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+   * @brief Calculate detailed balance, convert to MD and merge as the old way
+   *
+   */
+  void calculate_polarization_angle_correction(const std::string &event_ws_0, const std::string &event_ws_2,
+                                               const std::string &corrected_qsample_name,
+                                               const std::string &corrected_qlab_name,
+                                               const std::string &corrected_qsample_merged_name) {
+    const std::string temp_event_ws0("PolarizationAngleTempEvent0");
+
+    // apply detailed balance and convert to MD for event workspace 1
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:352]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+    alg.setProperty("OutputWorkspace", "ExpectToFail");
+
+    // Expect to fail due to missing temperature sample log
+    TS_ASSERT_THROWS_ANYTHING(alg.execute());
+  }
+
+  //---------------------------------------------------------------------------------------------
+  /**
+   * @brief Test apply detailed balance to 1 run
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:67]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+    convert_alg.setPropertyValue("InputWorkspace", event_ws_name);
+    convert_alg.setPropertyValue("OutputWorkspace", md_ws_name);
+    convert_alg.setPropertyValue("QDimensions", q_dimensions);
+    if (q3dframe != "")
+      convert_alg.setPropertyValue("Q3DFrames", q3dframe);
+    convert_alg.execute();
+  }
+
+  /**
+   * @brief Calculate detailed balance, convert to MD and merge as the old way
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:343]: wrong name

File: Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  if (dim_error != "") {
+    output["InputWorkspace"] = dim_error;
+  }
+
+  return output;
+}
+
+//---------------------------------------------------------------------------------------------
+/**
+ * @brief Apply detailed balance to each MDEvent in MDEventWorkspace
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp:142]: wrong name

File: Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+  auto precisionrange = std::make_shared<BoundedValidator<double>>();
+  precisionrange->setLower(0.);
+  precisionrange->setUpper(1.);
+  declareProperty("Precision", 1., precisionrange,
+                  "Precision (between 0 and 1). Any event whose cosine of 2 of its schaf angle less than this "
+                  "precision will be ignored.");
+
+  declareProperty(
+      std::make_unique<WorkspaceProperty<API::IMDEventWorkspace>>("OutputWorkspace", "", Kernel::Direction::Output),
+      "The output MDEventWorkspace with detailed balance applied");
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp:75]: no detailed balance

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/PolarizationAngleCorrectionMD.h [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+private:
+  /// Initialize the proeprties
+  void init() override;
+  /// Run the algorithm
+  void exec() override;
+  /// Validate inputs
+  std::map<std::string, std::string> validateInputs() override;
+
+  /// Apply detailed balance to each MDEvent
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/inc/MantidMDAlgorithms/PolarizationAngleCorrectionMD.h:39]: wrong comment

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+   */
+  void test_1run_qlab() {
+    // Check whether the MD to test does exist
+    auto singlemd = Mantid::API::AnalysisDataService::Instance().retrieve(mQLabWorkspaceName);
+    TS_ASSERT(singlemd);
+
+    // specify the output
+    std::string outputname("PolarizationAngleSingleQlabTest");
+
+    // Calculate detailed balance for the single MDEventWorkspace
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:77]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+
+  void test_1run_qsample() {
+    // Check whether the MD to test does exist
+    auto singlemd = Mantid::API::AnalysisDataService::Instance().retrieve(mQLabWorkspaceName);
+    TS_ASSERT(singlemd);
+
+    // specify the output
+    std::string outputname("PolarizationAngleSingleQsampleTest");
+
+    // Calculate detailed balance for the single MDEventWorkspace
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:107]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+
+    bool equals = compareMDEvents(outputname, mGoldCorrectedQSampleWSName, false);
+    TS_ASSERT(equals);
+
+    // Clean up
+    cleanWorkspace(outputname, false);
+  }
+
+  /**
+   * @brief Test applying detailed balance to 2 merged runs in Q_sample
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:129]: wrong name

File: Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h (hunk 6 above)
Comment (AndreiSavici) [inline Framework/MDAlgorithms/test/PolarizationAngleCorrectionMDTest.h:343]: wrong name

**Review by AndreiSavici — CHANGES_REQUESTED** [review]
Works as expected, but need simple changes as described.

File: Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+  declareProperty(
+      std::make_unique<WorkspaceProperty<API::IMDEventWorkspace>>("InputWorkspace", "", Kernel::Direction::Input),
+      "An input MDEventWorkspace.  Must be in Q_sample/Q_lab frame.  Must have an axis as DeltaE");
+
+  auto anglerange = std::make_shared<BoundedValidator<double>>();
+  anglerange->setLower(-180.);
+  anglerange->setUpper(180.);
+  declareProperty("PolarizationAngle", 0., anglerange,
+                  "An in-plane polarization angle PAP_APA​, between -180 and 180 degrees");
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/MDAlgorithms/src/PolarizationAngleCorrectionMD.cpp:64]: copy & paste error

**Review by AndreiSavici — APPROVED** [review]

**Pull request merged**